File: backend/src/utils/validators.py
"""
Platform-specific URL validators for video submission.
"""

from urllib.parse import urlparse, parse_qs

from src.models.video import VideoPlatform


def validate_platform_url(url: str, platform: VideoPlatform) -> bool:
    """
    Validate URL format based on detected platform.

    Args:
        url: URL to validate
        platform: Detected platform

    Returns:
        True if URL is valid for platform
    """
    parsed = urlparse(url)

    if not parsed.scheme or parsed.scheme not in ["http", "https"]:
        return False

    if platform == VideoPlatform.YOUTUBE:
        return _validate_youtube_url(parsed)
    if platform == VideoPlatform.VIMEO:
        return _validate_vimeo_url(parsed)
    if platform == VideoPlatform.DIRECT:
        return _validate_direct_url(parsed)

    return False


def _validate_youtube_url(parsed) -> bool:
    """Validate YouTube URL format."""
    domain = parsed.netloc.replace("www.", "").lower()

    if domain == "youtu.be":
        # Short URL format: youtu.be/VIDEO_ID
        return bool(parsed.path and parsed.path != "/")

    if domain in ["youtube.com", "m.youtube.com"]:
        # Standard format: youtube.com/watch?v=VIDEO_ID
        if parsed.path == "/watch":
            query_params = parse_qs(parsed.query)
            return "v" in query_params and bool(query_params["v"][0])

        # Embed format: youtube.com/embed/VIDEO_ID
        if parsed.path.startswith("/embed/"):
            return len(parsed.path.split("/")) >= 3

    return False


def _validate_vimeo_url(parsed) -> bool:
    """Validate Vimeo URL format."""
    domain = parsed.netloc.replace("www.", "").lower()

    if domain == "vimeo.com":
        # Format: vimeo.com/VIDEO_ID
        path_parts = [p for p in parsed.path.split("/") if p]
        return len(path_parts) >= 1 and path_parts[0].isdigit()

    if domain == "player.vimeo.com":
        # Embed format: player.vimeo.com/video/VIDEO_ID
        path_parts = [p for p in parsed.path.split("/") if p]
        return len(path_parts) >= 2 and path_parts[0] == "video" and path_parts[1].isdigit()

    return False


def _validate_direct_url(parsed) -> bool:
    """Validate direct video URL format."""
    # Must have valid domain
    if not parsed.netloc:
        return False

    # Check for video file extensions
    video_extensions = [".mp4", ".avi", ".mov", ".mkv", ".webm", ".flv", ".m4v"]
    if any(parsed.path.lower().endswith(ext) for ext in video_extensions):
        return True

    # Allow URLs with video-like paths
    video_path_indicators = ["video", "media", "stream", "content"]
    path_lower = parsed.path.lower()
    return any(indicator in path_lower for indicator in video_path_indicators)

File: frontend/src/utils/validators.ts
/**
 * URL validation utilities.
 */

/**
 * Check if a string is a valid URL.
 */
export function isValidUrl(url: string): boolean {
  try {
    const parsed = new URL(url);
    return parsed.protocol === 'http:' || parsed.protocol === 'https:';
  } catch {
    return false;
  }
}

/**
 * Check if URL is a YouTube video.
 */
export function isYouTubeUrl(url: string): boolean {
  try {
    const parsed = new URL(url);
    return (
      parsed.hostname === 'www.youtube.com' ||
      parsed.hostname === 'youtube.com' ||
      parsed.hostname === 'youtu.be' ||
      parsed.hostname === 'm.youtube.com'
    );
  } catch {
    return false;
  }
}

/**
 * Check if URL is a Vimeo video.
 */
export function isVimeoUrl(url: string): boolean {
  try {
    const parsed = new URL(url);
    return parsed.hostname === 'vimeo.com' || parsed.hostname === 'www.vimeo.com';
  } catch {
    return false;
  }
}

/**
 * Check if URL appears to be a direct video link.
 */
export function isDirectVideoUrl(url: string): boolean {
  try {
    const parsed = new URL(url);
    const videoExtensions = ['.mp4', '.avi', '.mov', '.mkv', '.webm', '.flv', '.m4v'];
    const path = parsed.pathname.toLowerCase();

    return videoExtensions.some((ext) => path.endsWith(ext));
  } catch {
    return false;
  }
}

/**
 * Extract video ID from YouTube URL.
 */
export function extractYouTubeId(url: string): string | null {
  try {
    const parsed = new URL(url);

    // youtu.be format
    if (parsed.hostname === 'youtu.be') {
      return parsed.pathname.slice(1);
    }

    // youtube.com/watch?v= format
    if (parsed.hostname.includes('youtube.com')) {
      return parsed.searchParams.get('v');
    }

    return null;
  } catch {
    return null;
  }
}

/**
 * Extract video ID from Vimeo URL.
 */
export function extractVimeoId(url: string): string | null {
  try {
    const parsed = new URL(url);
    const match = parsed.pathname.match(/\/(\d+)/);
    return match?.[1] ?? null;
  } catch {
    return null;
  }
}

/**
 * Validate and normalize video URL.
 */
export function validateVideoUrl(url: string): {
  isValid: boolean;
  platform?: 'youtube' | 'vimeo' | 'direct';
  videoId?: string;
  error?: string;
} {
  if (!url.trim()) {
    return { isValid: false, error: 'URL is required' };
  }

  if (!isValidUrl(url)) {
    return { isValid: false, error: 'Invalid URL format' };
  }

  if (isYouTubeUrl(url)) {
    const videoId = extractYouTubeId(url);
    if (!videoId) {
      return { isValid: false, error: 'Could not extract YouTube video ID' };
    }
    return { isValid: true, platform: 'youtube', videoId };
  }

  if (isVimeoUrl(url)) {
    const videoId = extractVimeoId(url);
    if (!videoId) {
      return { isValid: false, error: 'Could not extract Vimeo video ID' };
    }
    return { isValid: true, platform: 'vimeo', videoId };
  }

  // Validate direct video URL
  if (isDirectVideoUrl(url)) {
    return { isValid: true, platform: 'direct' };
  }

  return { isValid: false, error: 'Unsupported video URL. Use YouTube, Vimeo, or direct video file URL.' };
}
